File: codingTesting/그리디_큰 수 만들기.cpp
/*
프로그래머스 문제
큰 수 만들기

https://programmers.co.kr/learn/courses/30/lessons/42883
*/
#include <string>
#include <algorithm>
#include <vector>

using namespace std;

string solution(string number, int k) {
    string answer(number.substr(k));
    
    for(int i = k - 1; i >= 0 ; i--){
        int answer_index = 0;
        while(answer_index < answer.size()){
            if(number[i] >= answer[answer_index]){
                swap(number[i], answer[answer_index]);
                answer_index++;
            }else{
                break;
            }
        }
    }
    
    return answer;
}

File: codingTesting/그리디_큰 수 만들기.py
'''
프로그래머스 문제
큰 수 만들기

https://programmers.co.kr/learn/courses/30/lessons/42883
'''
def solution(number, k):
    stack = []
    
    for num in number:
        while len(stack) > 0 and num > stack[-1] and k > 0:
            stack.pop()
            k -= 1
        
        stack.append(num)
    
    if k > 0:
        stack = stack[:-k]
    
    answer = ''.join(stack)
    return answer
